### ISSUE TITLE
Commit signature not detected

### ISSUE THREAD

**Issue by jmah** [not-an-aardvark/lucky-commit#25]
My commit signing flow with X509 signatures results in the commit object containing:

```
gpgsig -----BEGIN SIGNED MESSAGE-----
 …
 -----END SIGNED MESSAGE-----
```

This is not detected by `lucky-commit` since it's explicitly looking for `-----END PGP SIGNATURE-----`.

**Comment by not-an-aardvark** [comment]
Out of curiosity, if you replace the hardcoded `BEGIN PGP SIGNATURE`/`END PGP SIGNATURE` markers in `lucky-commit`'s source code with `BEGIN SIGNED MESSAGE`/`END SIGNED MESSAGE` and then run it, does the resulting commit still pass signature verification? I'd be fine with adding support for more markers, but I haven't tested whether `lucky-commit`'s current approach of "adding whitespace to the signature" still works with signature formats other than PGP.

**Comment by jmah** [comment]
Just tried… unfortunately it fails validation.

```
commit 123423df0e050bf2e739c3be0efce92d53020ec4 (HEAD -> main)
Signer: unknown subject (unknown fingerprint) ❌ (signature invalid)
```

**Comment by not-an-aardvark** [comment]
It's possible that there are also other degrees of freedom in verifying an X509 signature (i.e. maybe it would work to add the whitespace to the signature in a different place from where `lucky-commit` currently puts it). Do you know of any public git commits offhand that have X509 signatures? I'm willing to try it out, but I've never used X509 signatures with git commits, so I don't have much test data.

(Apparently there are also [ssh signatures for git commits](https://github.com/git/git/blob/083e01275b81eb6105247ec522adf280d99f6b56/gpg-interface.c#L94) now, which seem to have yet another signature marker. I might need to investigate how git finds the end of a signature and replicate that, rather than just searching for hardcoded strings within the signature.)

**Comment by jmah** [comment]
Curiously I get a valid signature putting whitespace anywhere _within_ the signature block (between the BEGIN and END lines) — at least the locations I tried (at the end of lines, in the middle, on new whitespace-only lines…).

Surprisingly it also validates when I put printing characters before the END marker, after the final `==`.

**Comment by not-an-aardvark** [comment]
> Just tried  [replacing the hardcoded `END PGP SIGNATURE` with `END SIGNED MESSAGE`]… unfortunately it fails validation.

That's odd -- when I tested this out just now, it seems like the signature ends up being detected as valid.

To double-check, when you tried this out, was the signature valid before you ran `lucky_commit`? (If you had previously run `lucky_commit` on the same commit before monkeypatching the source code, then it would have amended the commit message, making the signature invalid. After doing that, the signature would still be invalid even after running the patched version of `lucky_commit` to amend the signature block instead. I think in order to test this out properly, it would be necessary to run it on a "fresh" commit.)

**Comment by not-an-aardvark** [comment]
As of https://github.com/not-an-aardvark/lucky-commit/commit/0000000916c011cf80aec842b5d76d479bd94a64, `lucky-commit` no longer searches for a hardcoded signature marker. Instead, it parses the header beginning with `gpgsig` and puts its padding at the end of the header (which should be the same place as before for PGP signatures, but should also work now for SSH/SMIME signatures).

From my local testing, SMIME signatures still seem to verify with the padding at that place. Feel free to let me know whether it works for you.

**Comment by Kingdutch** [comment]
First of all thanks for a really cool project! I saw someone use it to sync commit hashes between a source repo and a dist repository so you can easily find the source of changes.

However, it looks like for SSH keys this doesn't quite work yet.

I have [a git workflow](https://github.com/Kingdutch/cucumber-linter-src/blob/0.1.x/.github/workflows/phar.yml) that signs [both tags](https://github.com/Kingdutch/cucumber-linter-src/blob/0.1.x/.github/workflows/phar.yml#L269) as well [as commits](https://github.com/Kingdutch/cucumber-linter-src/blob/0.1.x/.github/workflows/phar.yml#L258). The commits are [modified by lucky_commit](https://github.com/Kingdutch/cucumber-linter-src/blob/0.1.x/.github/workflows/phar.yml#L259). Both use [the same config](https://github.com/Kingdutch/cucumber-linter-src/blob/0.1.x/.github/workflows/phar.yml#L249).

In the [commit history of the results](https://github.com/Kingdutch/cucumber-linter/commits?author=cucumber-linter-bot) I can see [the tag without modification signed correctly](https://github.com/Kingdutch/cucumber-linter/commit/27450c676d6fe803ad791a1f3961be0165398234) but the [regular commit with modification](https://github.com/Kingdutch/cucumber-linter/commit/dfe7b8473a53b8146361b2b006dfd934560dab7e) unfortunately didn't.

**Comment by not-an-aardvark** [comment]
> First of all thanks for a really cool project! I saw someone use it to sync commit hashes between a source repo and a dist repository so you can easily find the source of changes.

I'm glad people have been finding `lucky-commit` useful! That said, I have some doubts that this is the best way to track the provenance of changes from a source repo. It seems like it would be much more straightforward to simply include the commit hash of the source project's commit in the commit message of the destination project's commit. That would avoid any issues in case the source repo has two commits with the same shorthash.

> However, it looks like for SSH keys this doesn't quite work yet.

This might be because the fix commit (https://github.com/not-an-aardvark/lucky-commit/commit/0000000916c011cf80aec842b5d76d479bd94a64) was never actually published to a release. (I'd been waiting for the OP to confirm that the fix worked, but that never happened.) I've published v2.2.2, so hopefully this should start working now.

**Comment by Kingdutch** [comment]
> That said, I have some doubts that this is the best way to track the provenance of changes from a source repo. It seems like it would be much more straightforward to simply include the commit hash of the source project's commit in the commit message of the destination project's commit. That would avoid any issues in case the source repo has two commits with the same shorthash.

You're probably technically right. Though I stole this method from [a much larger project](https://github.com/phpstan/phpstan) which doesn't seem to have run into problems yet, so I'm not too worried for now.

The automation that I borrowed from them also pushes a changelog into the commit body, but I must say that being able to list the shorthashes side by side in the GitHub UI is very useful (or even just being able to copy from one and paste into a `git log` for the other).

> This might be because the fix commit (https://github.com/not-an-aardvark/lucky-commit/commit/0000000916c011cf80aec842b5d76d479bd94a64) was never actually published to a release. (I'd been waiting for the OP to confirm that the fix worked, but that never happened.) I've just published v2.2.2, so hopefully this should start working now.

That would definitely explain it ^^ I just pushed an empty commit into the source directory to trigger a rebuild and the commit came out verified on the other end! 🥳 

<img width="1232" alt="image" src="https://github.com/not-an-aardvark/lucky-commit/assets/327697/5b61a31d-303f-4b61-806d-fb64d68f4e20">
